Add documentation for the new configuration options for Kafka external listeners (#1352)
strimzi/strimzi-kafka-operator: documentation/book/con-kafka-listeners.adoc

When exposing Kafka using {OpenShiftName} `Routes`, TLS encryption is always used.

> By default, the route hosts will be assigned by {OpenShiftName}.

Author: The Routes exist only on OpenShift - hence the `{OpenShiftName}`. `{ProductPlatformName}` would add Kubernetes as well her which we don't want.

You can override the auto assigned route hosts by specifying the requested host in the `overrides` property.
{ProductName} will not perform any validation that the requested hosts are available.
You have to make sure that the hosts are free and can be used.

.An example where the external listener is configured with overrides for {OpenShiftName} route hosts
[source,yaml,subs="attributes+"]
----
# ...
listeners:
  external:
    type: routes
    authentication:
      type: tls
    overrides:
      bootstrap:
        host: bootstrap.myrouter.com
      brokers:
      - broker: 0
        host: broker-0.myrouter.com
      - broker: 1
        host: broker-1.myrouter.com
      - broker: 2
        host: broker-2.myrouter.com
# ...
----

> For more information on using `Routes` to access Kafka, see xref:proc-accessing-kafka-using-routes-{context}[].

Reviewer: As already done for the following section about nodeports, should be say the same about `bootstrap`.
I am referring to the paragraph : 

> Additionally, you can specify the name of the bootstrap service.
This name will be added to the broker certificates and can be used for TLS hostname verification.

Author: That is not in the NodePort section. There is a section at the end called `Customising advertised addresses on external listeners` which applies to all external listeners.

Author: I hope the clarification sentence which I'm adding as part of the third comment will make it more clear.

In some situations this might not be sufficient, because the infrastructure where you run {ProductName} might not provide the right hostname on port under which Kafka can be accessed.
You can customize the advertised hostname and port in the `overrides` property of the external listener.
{ProductName} will automatically configure the advertised address in the Kafka brokers and add it to the broker certificates, so that it can be used for TLS hostname verification.

.An example where the external listener is configured with overrides for advertised addresses
[source,yaml,subs="attributes+"]
----
# ...
listeners:
  external:
    type: routes
    authentication:
      type: tls
    overrides:
      brokers:
      - broker: 0
        advertisedHost: example.hostname.0
        advertisedPort: 12340
      - broker: 1
        advertisedHost: example.hostname.1
        advertisedPort: 12341
      - broker: 2
        advertisedHost: example.hostname.2
        advertisedPort: 12342
# ...
----

Additionally, you can specify the name of the bootstrap service.

> This name will be added to the broker certificates and can be used for TLS hostname verification.

Reviewer: because most of the above examples are with `routes` we could say that the `bootstrap.address` the customisation is possible for all the external listener types.

> .An example where the external listener is configured with overrides for advertised addresses

Reviewer: is it wrong right? the example is about overrides the bootstrap service.

        nodePort: 32100
      brokers:
      - broker: 0
        nodePort: 32000
      - broker: 1
        nodePort: 32001
      - broker: 2
        nodePort: 32002
# ...
----

For more information on using node ports to access Kafka, see xref:proc-accessing-kafka-using-nodeports-{context}[].

> .Customising advertised addresses on external listeners

Reviewer 2: ```suggestion
.Customizing advertised addresses on external listeners
```

> By default, {ProductName} will try to automatically determine the hostnames and ports which the Kafka cluster should advertise to its clients.

Reviewer 2: ```suggestion
By default, {ProductName} tries to automatically determine the hostnames and ports that your Kafka cluster advertises to its clients.
```

> In some situations this might not be sufficient, because the infrastructure where you run {ProductName} might not provide the right hostname on port under which Kafka can be accessed.

Reviewer 2: ```suggestion
This is not sufficient in all situations, because the infrastructure on which {ProductName} is running might not provide the right hostname or port through which Kafka can be accessed.
```

Reviewer 2: _Begin a new paragraph._

You can customize the advertised hostname and port in the `overrides` property of the external listener.

> {ProductName} will automatically configure the advertised address in the Kafka brokers and add it to the broker certificates, so that it can be used for TLS hostname verification.

Reviewer 2: ```suggestion
{ProductName} will then automatically configure the advertised address in the Kafka brokers and add it to the broker certificates so it can be used for TLS hostname verification.
```

Overriding the advertised host and ports is available for all types of external listeners.

> .An example where the external listener is configured with overrides for advertised addresses

Reviewer 2: ```suggestion
.Example of an external listener configured with overrides for advertised addresses
```

[source,yaml,subs="attributes+"]
----
# ...
listeners:
  external:
    type: routes
    authentication:
      type: tls
    overrides:
      brokers:
      - broker: 0
        advertisedHost: example.hostname.0
        advertisedPort: 12340
      - broker: 1
        advertisedHost: example.hostname.1
        advertisedPort: 12341
      - broker: 2
        advertisedHost: example.hostname.2
        advertisedPort: 12342
# ...
----

Additionally, you can specify the name of the bootstrap service.
This name will be added to the broker certificates and can be used for TLS hostname verification.
Adding the additional bootstrap address is available for all types of external listeners.

> .An example where the external listener is configured with additional bootstrap address

Reviewer 2: ```suggestion
.Example of an external listener configured with an additional bootstrap address
```
